cpp/src/arrow/compute/exec/hash_join_benchmark.cc

```diff
@@ -0,0 +1,406 @@
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_KeyTypes, "{fixed_size_binary(32)}",
+                  {fixed_size_binary(32)})
+    ->ArgNames(keytypes_argnames)
+    ->RangeMultiplier(4)
+    ->Range(1, 4 * 1024);
+
+std::vector<std::string> selectivity_argnames = {"Selectivity", "HashTable krows"};
+std::vector<std::vector<int64_t>> selectivity_args = {
+    benchmark::CreateDenseRange(0, 100, 20), benchmark::CreateRange(1, 4 * 1024, 8)};
+
+BENCHMARK_CAPTURE(BM_HashJoinBasic_Selectivity, "{int32}", {int32()})
+    ->ArgNames(selectivity_argnames)
+    ->ArgsProduct(selectivity_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_Selectivity, "{int64}", {int64()})
+    ->ArgNames(selectivity_argnames)
+    ->ArgsProduct(selectivity_args);
+
+// Joins on UTF8 are currently really slow, so anything above 512 doesn't finished within
+// a reasonable amount of time.
+BENCHMARK_CAPTURE(BM_HashJoinBasic_Selectivity, "{utf8}", {utf8()})
+    ->ArgNames(selectivity_argnames)
+    ->ArgsProduct({benchmark::CreateDenseRange(0, 100, 20),
+                   benchmark::CreateRange(1, 64, 8)});
+
+BENCHMARK_CAPTURE(BM_HashJoinBasic_Selectivity, "{fixed_size_binary(16)}",
+                  {fixed_size_binary(16)})
+    ->ArgNames(selectivity_argnames)
+    ->ArgsProduct(selectivity_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_Selectivity, "{fixed_size_binary(32)}",
+                  {fixed_size_binary(32)})
+    ->ArgNames(selectivity_argnames)
+    ->ArgsProduct(selectivity_args);
+
+std::vector<std::string> jointype_argnames = {"Selectivity", "HashTable krows"};
+std::vector<std::vector<int64_t>> jointype_args = {
+    benchmark::CreateDenseRange(0, 100, 20), benchmark::CreateRange(1, 4 * 1024, 8)};
+
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Inner", JoinType::INNER)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Left Semi", JoinType::LEFT_SEMI)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Right Semi", JoinType::RIGHT_SEMI)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Left Anti", JoinType::LEFT_ANTI)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Right Anti", JoinType::RIGHT_ANTI)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Left Outer", JoinType::LEFT_OUTER)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Right Outer", JoinType::RIGHT_OUTER)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
+BENCHMARK_CAPTURE(BM_HashJoinBasic_JoinType, "Full Outer", JoinType::FULL_OUTER)
+    ->ArgNames(jointype_argnames)
+    ->ArgsProduct(jointype_args);
```

Comment:
I'll admit my ignorance here, but is it useful to benchmark all join types or are some just going to yield the same performance of another (because of symmetry)?

Reply:
Yes, because the build side is always on the right and the probe side is always on the left, so it's not actually symmetrical.